src/java.base/share/classes/jdk/internal/util/Architecture.java

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright (c) 2018, 2023, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package jdk.internal.util;
+
+import jdk.internal.vm.annotation.ForceInline;
+
+/**
+ * System architecture enum values.
+ * Each architecture has a matching {@code public static boolean isXXX()} method
+ * that is true when running on that architecture.
+ * The enum values correspond to architectures as named in the build system.
+ * See the values of `OPENJDK_TARGET_ARCH_OSARCH`.
+ * The order and ordinal values must match the values defined in OperatingSystemProps.java.template.
+ */
+public enum Architecture {
+    X64(),        // Represents AMD64 and X86_64
+    X86(),
+    IA64(),
+    ARM(),
+    AARCH64(),
+    RISCV64(),
+    S390X(),
+    PPC64LE(),
+    ;
+
+    // Cache a copy of the array for lightweight indexing
+    private static final Architecture[] archValues = Architecture.values();
```

Comment:
This needs to be annotated with `@jdk.internal.vm.annotation.Stable` for `Architecture.current()` to be constant foldable:
```suggestion
    private static final @Stable Architecture[] archValues = Architecture.values();
```

Reply:
Even if it's `static` *and* `final`? I thought `@Stable` exists to "...process non-null stable fields (final or otherwise) in a similar manner to static final fields with respect to promoting the field's value to a constant", implying that `static final` fields already have this property.

Reply:
The use here is not particularly performance sensitive. So it doesn't provide much value.
In this case, it allows each array element to be considered stable.

From the \@Stable javadoc:
```
 * Fields which are declared {@code final} may also be annotated as stable.
 * Since final fields already behave as stable values, such an annotation
 * conveys no additional information regarding change of the field's value, but
 * still conveys information regarding change of additional components values if
 * the type of the field is an array type (as described above).
```